internal/services/cosmos/cosmosdb_sql_role_definition_resource.go

```diff
@@ -0,0 +1,277 @@
+package cosmos
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/cosmos-db/mgmt/2021-10-15/documentdb"
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform-provider-azurerm/helpers/azure"
+	"github.com/hashicorp/terraform-provider-azurerm/helpers/tf"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/cosmos/parse"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/cosmos/validate"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/timeouts"
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+)
+
+func resourceCosmosDbSQLRoleDefinition() *pluginsdk.Resource {
+	return &pluginsdk.Resource{
+		Create: resourceCosmosDbSQLRoleDefinitionCreate,
+		Read:   resourceCosmosDbSQLRoleDefinitionRead,
+		Update: resourceCosmosDbSQLRoleDefinitionUpdate,
+		Delete: resourceCosmosDbSQLRoleDefinitionDelete,
+
+		Timeouts: &pluginsdk.ResourceTimeout{
+			Create: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Read:   pluginsdk.DefaultTimeout(5 * time.Minute),
+			Update: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Delete: pluginsdk.DefaultTimeout(30 * time.Minute),
+		},
+
+		Importer: pluginsdk.ImporterValidatingResourceId(func(id string) error {
+			_, err := parse.SqlRoleDefinitionID(id)
+			return err
+		}),
+
+		Schema: map[string]*pluginsdk.Schema{
+			"role_definition_id": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IsUUID,
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"account_name": {
+				Type:         pluginsdk.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.CosmosAccountName,
+			},
+
+			"type": {
+				Type:     pluginsdk.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  string(documentdb.RoleDefinitionTypeCustomRole),
+				ValidateFunc: validation.StringInSlice([]string{
+					string(documentdb.RoleDefinitionTypeBuiltInRole),
+					string(documentdb.RoleDefinitionTypeCustomRole),
+				}, false),
+			},
+
+			"assignable_scopes": {
+				Type:     pluginsdk.TypeSet,
+				Required: true,
+				Elem: &pluginsdk.Schema{
+					Type:         pluginsdk.TypeString,
+					ValidateFunc: validation.StringIsNotEmpty,
+				},
+			},
+
+			"name": {
+				Type:         pluginsdk.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"permissions": {
+				Type:     pluginsdk.TypeSet,
+				Required: true,
+				Elem: &pluginsdk.Resource{
+					Schema: map[string]*pluginsdk.Schema{
+						"data_actions": {
```

Comment:
Is this a block for a reason? or would it make more sense to pop this up a level into lets say `data_actions_allowed` ?

Reply:
@katbyte , there is an existing resource [azurerm_role_definition](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/role_definition#permissions) that is similar with this RP. So as @aristosvo above mentioned, it is better to build this cosmosdb sql role definition in a similar way. Then user could be easier to use and understand.